test/parallel/test-net-connect-buffer.js

```diff
@@ -43,21 +43,13 @@ const tcp = net.Server(common.mustCall((s) => {
 tcp.listen(0, common.mustCall(function() {
   const socket = net.Stream({ highWaterMark: 0 });
 
-  let connected = false;
-  assert.strictEqual(socket.pending, true);
-  socket.connect(this.address().port, common.mustCall(() => connected = true));
-
-  assert.strictEqual(socket.pending, true);
-  assert.strictEqual(socket.connecting, true);
-  assert.strictEqual(socket.readyState, 'opening');
-
   // Make sure that anything besides a buffer or a string throws.
-  common.expectsError(() => socket.write(null),
-                      {
-                        code: 'ERR_STREAM_NULL_VALUES',
-                        type: TypeError,
-                        message: 'May not write null values to stream'
-                      });
+  socket.write(null)
+  socket.on('error', common.expectsError({
+    code: 'ERR_STREAM_NULL_VALUES',
+    type: TypeError
+  }));
```

Comment:
Why does this change happen?

Reply:
@mcollina: Resolved it after adding docs and test?

Reply:
Because we no longer emit the error in the same tick, thus it doesn't throw synchronously.

Reply:
Let me dig into this again and I'll try to give you a better explanation.

Reply:
Agreed, what change in your PR is triggering this? I think throwing sync in case of validation error is ok.

Reply:
there is an "incorrect" behaviour in `errorOrDestroy` in that if `autoDestroy: false` it will emit the `'error'` synchronously instead of next tick. Since this PR changes `autoDestroy: true` the `'error'` is emitted correctly on next tick, which requires the test to be modified.

Reply:
> I think throwing sync in case of validation error is ok.

I guess that's fine. But it is inconsistent with how it's done everywhere else?

Reply:
I don't really understand this. How come it moves from being thrown to being emitted as an error?

Reply:
When emitting 'error' without an error listener it gets thrown.

Reply:
https://github.com/nodejs/node/blob/master/lib/events.js#L168